homework9/hw3.py

```diff
@@ -0,0 +1,40 @@
+"""
+Write a function that takes directory path, a file extension and an optional tokenizer.
+It will count lines in all files with that extension if there are no tokenizer.
+If a the tokenizer is not none, it will count tokens.
+
+# For dir with two files from hw1.py:
+# >>> universal_file_counter(test_dir, "txt")
+# 6
+# >>> universal_file_counter(test_dir, "txt", str.split)
+# 6
+
+"""
+import os
+from pathlib import Path
+from typing import Callable, Optional
+
+
+def tokenizer_processing(line, tokenizer):
+    inner_counter = 0
+    if tokenizer is None:
+        inner_counter += 1
+    else:
+        if type(tokenizer) is list:
+            inner_counter = len([1 for element in line.split(*tokenizer)])
+        else:
+            inner_counter = len([1 for element in tokenizer(line)])
+    return inner_counter
```

Comment:
Т.к. внутри нет цикла, возиться с увеличением счётчика смысла нет. Значение можно либо просто присваивать либо вообще сразу возвращать. Хотя функции с множеством return обычно сложнее понимать.

Можно сразу считать длину самой коллекции или генератора без ручной генерации списка из единичек.

[line.split()](https://docs.python.org/3/library/stdtypes.html?highlight=str.split#str.split) принимает первым параметром строку-разделитель и вторым максимальное количество разделений. Если передать туда в качестве аргументов кучу функций-токенайзеров вылетит ошибка

```suggestion
    if tokenizer is None:
        return 1
    else:
        if type(tokenizer) is list:
            return sum([len(line.split(entry)) for entry in tokenizer])
        else:
            return len(tokenizer(line))
    return 0
```